File: backend/app/ai/router.py
"""
AI Module — Router
"""
from datetime import date

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel, Field

from app.ai.service import AIAdvisorService, get_ai_advisor_service

router = APIRouter()

ALLOWED_SCENARIOS = {
    "price_increase",
    "hire_cleaner",
    "expense_reduction",
    "minimum_stay",
}


class ChatRequest(BaseModel):
    question: str = Field(..., min_length=1, max_length=500)
    year: int | None = None


class ScenarioRequest(BaseModel):
    scenario: str
    params: dict = {}
    year: int | None = None


@router.get("/analyze")
async def analyze_financial_performance(
    service: AIAdvisorService = Depends(get_ai_advisor_service),
):
    """AI-driven financial performance analysis with recommendations."""
    return await service.analyze_financial_performance()


@router.get("/advisor")
async def get_advisor_report(
    year: int | None = Query(None, ge=2020, le=2100),
    start_date: date | None = Query(None),
    end_date: date | None = Query(None),
    service: AIAdvisorService = Depends(get_ai_advisor_service),
):
    """
    Comprehensive AI advisor dashboard: executive summary, business
    health, priority actions, opportunities, lost revenue, risks,
    property reviews, 30-day forecast, achievements, goals.

    Either `year` (defaults to the current year) or a `start_date`/`end_date`
    pair selects the reporting period.
    """
    if (start_date is None) != (end_date is None):
        raise HTTPException(
            status_code=422,
            detail="start_date and end_date must be provided together",
        )
    return await service.generate_advisor_report(year, start_date, end_date)


@router.post("/chat")
async def chat_with_advisor(
    data: ChatRequest,
    service: AIAdvisorService = Depends(get_ai_advisor_service),
):
    """Ask HostWise AI a natural-language question about your portfolio."""
    year = data.year or date.today().year
    return await service.answer_question(data.question, year)


@router.post("/test-connection")
async def test_ai_connection(
    service: AIAdvisorService = Depends(get_ai_advisor_service),
):
    """Validate the configured BYOK LLM connection."""
    return await service.test_llm_connection()


@router.post("/scenario")
async def run_scenario(
    data: ScenarioRequest,
    service: AIAdvisorService = Depends(get_ai_advisor_service),
):
    """Simulate a 'what-if' scenario and estimate its financial impact."""
    if data.scenario not in ALLOWED_SCENARIOS:
        raise HTTPException(
            status_code=400,
            detail=f"Unknown scenario. Allowed: {sorted(ALLOWED_SCENARIOS)}",
        )
    year = data.year or date.today().year
    return await service.simulate_scenario(data.scenario, data.params, year)

File: app/backend/src/ai/router.rs
//! AI HTTP routes, mirroring `backend/app/ai/router.py`.
//!
//! GET  /api/v1/ai/analyze
//! GET  /api/v1/ai/advisor?year=|start_date=&end_date=
//! POST /api/v1/ai/chat  { question, year? }
//! POST /api/v1/ai/test-connection
//! POST /api/v1/ai/scenario { scenario, params?, year? }

use axum::extract::{Query, State};
use axum::routing::{get, post};
use axum::{Json, Router};
use chrono::Datelike;
use serde::Deserialize;
use serde_json::Value;

use crate::ai::service::AiAdvisorService;
use crate::core::error::AppError;
use crate::core::state::AppState;

const ALLOWED_SCENARIOS: &[&str] = &[
    "price_increase",
    "hire_cleaner",
    "expense_reduction",
    "minimum_stay",
];

#[derive(Debug, Deserialize)]
struct AdvisorQuery {
    year: Option<i32>,
    start_date: Option<String>,
    end_date: Option<String>,
}

#[derive(Debug, Deserialize)]
struct ChatRequest {
    question: String,
    year: Option<i32>,
}

#[derive(Debug, Deserialize)]
struct ScenarioRequest {
    scenario: String,
    #[serde(default)]
    params: Value,
    year: Option<i32>,
}

pub fn build_router() -> Router<AppState> {
    Router::new()
        .route("/analyze", get(analyze))
        .route("/advisor", get(advisor))
        .route("/chat", post(chat))
        .route("/test-connection", post(test_connection))
        .route("/scenario", post(scenario))
}

fn svc(state: AppState) -> AiAdvisorService {
    AiAdvisorService::new(state)
}

async fn analyze(State(state): State<AppState>) -> Result<Json<Value>, AppError> {
    let result = svc(state).analyze_financial_performance().await?;
    Ok(Json(result))
}

async fn advisor(
    State(state): State<AppState>,
    Query(q): Query<AdvisorQuery>,
) -> Result<Json<Value>, AppError> {
    if (q.start_date.is_some()) != (q.end_date.is_some()) {
        return Err(AppError::Validation(
            "start_date and end_date must be provided together".into(),
        ));
    }
    let result = svc(state)
        .generate_advisor_report(q.year, q.start_date.as_deref(), q.end_date.as_deref())
        .await?;
    Ok(Json(result))
}

async fn chat(
    State(state): State<AppState>,
    Json(req): Json<ChatRequest>,
) -> Result<Json<Value>, AppError> {
    if req.question.trim().is_empty() || req.question.len() > 500 {
        return Err(AppError::Validation(
            "question must be 1-500 characters".into(),
        ));
    }
    let year = req.year.unwrap_or_else(|| chrono::Local::now().year());
    let result = svc(state)
        .answer_question(req.question.trim(), year)
        .await?;
    Ok(Json(result))
}

async fn test_connection(State(state): State<AppState>) -> Result<Json<Value>, AppError> {
    let result = svc(state).test_llm_connection().await?;
    Ok(Json(result))
}

async fn scenario(
    State(state): State<AppState>,
    Json(req): Json<ScenarioRequest>,
) -> Result<Json<Value>, AppError> {
    if !ALLOWED_SCENARIOS.contains(&req.scenario.as_str()) {
        return Err(AppError::Validation(format!(
            "Unknown scenario. Allowed: {}",
            ALLOWED_SCENARIOS.join(", ")
        )));
    }
    let year = req.year.unwrap_or_else(|| chrono::Local::now().year());
    let result = svc(state)
        .simulate_scenario(&req.scenario, &req.params, year)
        .await?;
    Ok(Json(result))
}
